pyscf_ipu/electron_repulsion/popcint/libcint.py

```diff
@@ -0,0 +1,506 @@
+
+def ipu_getints4c(intor_name, atm, bas, env, N, shls_slice=None, comp=1,
+            aosym='s1', ao_loc=None, cintopt=None, out=None, which_integral=-1):
+    natm = atm.shape[0]
+    nbas = bas.shape[0]
+
+    shls_slice = (0, nbas, 0, nbas, 0, nbas, 0, nbas)
+    
+    shape = [comp, N, N, N, N] 
+    
+    drv = libcgto.GTOnr2e_fill_drv
+    fill = getattr(libcgto, 'GTOnr2e_fill_'+aosym)
+    out = numpy.ndarray(shape, buffer=out)
+
+    # type 
+    float32 = "#define dtype float" in open("libcint.c", "r").read()
+    if float32: 
+        out = out.astype(np.float32)
+        env = env.astype(np.float32)
+
+
+    from tessellate_ipu import create_ipu_tile_primitive, ipu_cycle_count, tile_map, tile_put_sharded, tile_put_replicated
+    vertex_filename  = osp.join(osp.dirname(__file__), "libcint.cpp")
+    grad = create_ipu_tile_primitive(
+            "Int2e" ,
+            "Int2e" ,
+            inputs=["mat", "shls_slice", "ao_loc", "atm", "bas", "env", "natm", "nbas", "which_integral", "comp"], 
+            outputs={"out": 0},
+            gp_filename=vertex_filename,
+            perf_estimate=100,
+    )
+
+    natm = atm.shape[0]
+    nbas = bas.shape[0]
+
+    prefix = 'GTO'
+
+    float32 = "#define dtype float" in open("libcint.c", "r").read()
+    if float32: 
+        out = out.astype(np.float32)
+        env = env.astype(np.float32)
+
+
+    out = tile_put_replicated(jnp.array(out, dtype=jnp.float32),   (1,)) 
+    shls_slice = tile_put_replicated(jnp.array(shls_slice, dtype=jnp.int32),   (1,)) 
+    ao_loc = tile_put_replicated(jnp.array(ao_loc, dtype=jnp.int32),   (1,)) 
+    atm = tile_put_replicated(jnp.array(atm, dtype=jnp.int32),   (1,)) 
+    bas = tile_put_replicated(jnp.array(bas, dtype=jnp.int32),   (1,)) 
+    env = tile_put_replicated(jnp.array(env, dtype=jnp.float32),   (1,)) 
+    natm = tile_put_replicated(jnp.array(natm, dtype=jnp.int32),   (1,)) 
+    nbas = tile_put_replicated(jnp.array(nbas, dtype=jnp.int32),   (1,)) 
+    comp = tile_put_replicated(jnp.array(comp, dtype=jnp.int32),   (1,)) 
+
+    which_integral = tile_put_replicated(np.array(which_integral, dtype=jnp.int32),   (1,)) 
+
+    value = tile_map(grad, out, shls_slice, ao_loc, atm, bas, env, natm, nbas, which_integral, comp)
+
+    out = value.array 
+
+    if comp == 1:
+        out = out[0]
+    return out
+
+
+
+def ipu_tile_map_getints4c(intor_name, atm, bas, env, N, shls_slice=None, comp=1,
+            aosym='s1', ao_loc=None, cintopt=None, out=None, which_integral=-1):
+    natm = atm.shape[0]
+    nbas = bas.shape[0]
+
+    shls_slice = (0, nbas, 0, nbas, 0, nbas, 0, nbas)
+    
+    shape = [comp, N, N, N, N] 
+    
+    drv = libcgto.GTOnr2e_fill_drv
+    fill = getattr(libcgto, 'GTOnr2e_fill_'+aosym)
+    out = numpy.ndarray(shape, buffer=out)
+
+    # type 
+    float32 = "#define dtype float" in open("libcint.c", "r").read()
+    if float32: 
+        out = out.astype(np.float32)
+        env = env.astype(np.float32)
```

Comment:
There's a lot of duplicated code here - and a lot of it is slow - so it feels as if there might be an opportunity to dedup.

Reply:
> There's a lot of duplicated code here - and a lot of it is slow - so it feels as if there might be an opportunity to dedup.

Sure, let's make those refactors. The heavy compute part is the 20k lines of C. Most of the above is the easy 1% optimizations I was hoping to off-load.